packages/compass-aggregations/src/modules/pipeline-builder/pipeline-ai.ts

```diff
@@ -0,0 +1,422 @@
+  Promise<void>,
+  | AIPipelineStartedAction
+  | AIPipelineFailedAction
+  | AIPipelineSucceededAction
+  | LoadNewPipelineAction
+> => {
+  return async (dispatch, getState, { atlasService, pipelineBuilder }) => {
+    const {
+      pipelineBuilder: {
+        aiPipeline: { aiPipelineFetchId: existingFetchId },
+      },
+      namespace,
+      dataService: { dataService },
+    } = getState();
+
+    if (aiPipelineFetchId !== -1) {
+      // Cancel the active request as this one will override.
+      abort(existingFetchId);
+    }
+
+    const abortController = new AbortController();
+    const { id: fetchId, signal } = getAbortSignal();
+
+    dispatch({
+      type: AIPipelineActionTypes.AIPipelineStarted,
+      fetchId,
+    });
+
+    let jsonResponse;
+    try {
+      const sampleDocuments =
+        (await dataService?.sample?.(
+          namespace,
+          {
+            query: {},
+            size: NUM_DOCUMENTS_TO_SAMPLE,
+          },
+          {
+            maxTimeMS: preferences.getPreferences().maxTimeMS,
+            promoteValues: false,
+          },
+          {
+            abortSignal: signal,
+          }
+        )) || [];
+      const schema = await getSimplifiedSchema(sampleDocuments);
+
+      const { collection: collectionName, database: databaseName } =
+        toNS(namespace);
+      jsonResponse = await atlasService.getAggregationFromUserInput({
+        signal: abortController.signal,
+        userInput,
+        collectionName,
+        databaseName,
+        schema,
+        // sampleDocuments, // For now we are not passing sample documents to the ai.
+      });
+    } catch (err: any) {
+      if (signal.aborted) {
+        // If we already aborted so we ignore the error.
+        return;
+      }
+      logFailed(err?.message);
+      // We're going to reset input state with this error, show the error in the
+      // toast instead
+      if (err.statusCode === 401) {
+        openToast('ai-unauthorized', {
+          variant: 'important',
+          title: 'Network Error',
+          description: 'Unauthorized',
+          timeout: 5000,
+        });
+      }
+      dispatch({
+        type: AIPipelineActionTypes.AIPipelineFailed,
+        errorMessage: err?.message,
+        networkErrorCode: err.statusCode,
+      });
+      return;
+    } finally {
+      // Remove the AbortController from the Map as we either finished
+      // waiting for the fetch or cancelled at this point.
+      cleanupAbortSignal(fetchId);
+    }
+
+    if (signal.aborted) {
+      log.info(
+        mongoLogId(1_001_000_227),
+        'AIPipeline',
+        'Cancelled ai pipeline request'
+      );
+      return;
+    }
+
+    let pipeline;
+    try {
+      if (!jsonResponse?.content?.aggregation?.pipeline) {
+        throw new Error(
+          'No pipeline returned. Please try again with a different prompt.'
+        );
+      }
+
+      // TODO(COMPASS-7105): Update the format the backend/ai model
```

Comment:
Linked the wrong Compass ticket here, should be [COMPASS-7100](https://jira.mongodb.org/projects/COMPASS/issues/COMPASS-7100)
Going to likely update this before merging based on when https://github.com/10gen/compass-mongodb-com/pull/89 is merged in.